src/server.rs

```diff
@@ -83,12 +81,37 @@ impl Conmon for ConmonServerImpl {
     }
 }
 
-// Use the single threaded runtime to save rss memory
-#[tokio::main(flavor = "current_thread")]
-async fn main() -> Result<(), Error> {
-    let (shutdown_tx, shutdown_rx) = tokio::sync::oneshot::channel();
+fn main() {
+    // First, initialize the server so we have access to the config pre-fork.
+    let server = ConmonServerImpl::new().unwrap();
+
+    // We need to fork as early as possible, especially before setting up tokio.
+    // if we don't, the child will have a strange thread space and we're at risk of deadlocking.
+    // We also have to treat the parent as the child (as described in [1]) to ensure we don't
+    // interrupt the child's execution
+    // 1: https://docs.rs/nix/0.23.0/nix/unistd/fn.fork.html#safety
+    match unsafe{fork()} {
+        Ok(ForkResult::Parent { child, .. }) => {
+            if let Some(path) = server.config().conmon_pidfile() {
+                debug!("Writing child pid to pidfile {}", path.display());
+                let child_str = format!("{}", child);
+                File::create(path)
+                    .unwrap()
```

Comment:
Then you can use `?` for a clean error instead of creating a core dump if we can't write the file.

Also similarly for the `fork()` invocation, it can just be `fork()?` and you don't need to match on `Ok/Err`.

Reply:
unfortunately we still need the match for the fork case because we need to branch between parent/child of the fork, but it looks much cleaner for the other cases

Reply:
Right, but you should be able drop the `Ok/Err` bits of that match by using `fork?`.